File: code-templates/matlab/ml/random_forest_template.m
% 随机森林 MATLAB 模板 (需要 Statistics and Machine Learning Toolbox)

function [model, metrics] = rf_template()
    % 加载数据
    % TODO: data = readtable('data.csv');
    rng(42);
    n = 500;
    X = randn(n, 4);
    y = 3*X(:,1) + 2*X(:,2).^2 - 1.5*X(:,3) + 0.5*X(:,4).*X(:,1) + 0.5*randn(n,1);

    % 划分
    cv = cvpartition(n, 'HoldOut', 0.2);
    X_train = X(training(cv), :);
    y_train = y(training(cv));
    X_test = X(test(cv), :);
    y_test = y(test(cv));

    % 标准化
    [X_train, mu, sigma] = zscore(X_train);
    X_test = (X_test - mu) ./ sigma;

    % 训练随机森林
    model = TreeBagger(300, X_train, y_train, ...
        'Method', 'regression', ...
        'MinLeafSize', 5, ...
        'NumPredictorsToSample', 'all', ...
        'OOBPrediction', 'on');

    % 预测
    y_pred = str2double(predict(model, X_test));

    % 评估
    r2 = 1 - sum((y_test - y_pred).^2) / sum((y_test - mean(y_test)).^2);
    mae = mean(abs(y_test - y_pred));
    rmse = sqrt(mean((y_test - y_pred).^2));

    fprintf('\n===== 随机森林 评估结果 =====\n');
    fprintf('R²  = %.4f\n', r2);
    fprintf('MAE = %.4f\n', mae);
    fprintf('RMSE = %.4f\n', rmse);

    % 特征重要性
    imp = model.OOBPermutedPredictorDeltaError;
    figure;
    barh(imp);
    xlabel('Importance');
    title('Random Forest Feature Importance');
    set(gca, 'YTickLabel', {'x1','x2','x3','x4'});

    metrics = struct('r2', r2, 'mae', mae, 'rmse', rmse);
end

File: code-templates/python/ml/random_forest_template.py
"""
随机森林 回归/分类模板
适用：表格数据、中等样本、需要可解释性
"""

import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
import matplotlib.pyplot as plt


def train_random_forest(X_train, X_test, y_train, y_test):
    """训练随机森林并评估"""
    model = RandomForestRegressor(
        n_estimators=300,
        max_depth=None,
        min_samples_split=5,
        min_samples_leaf=2,
        max_features='sqrt',
        random_state=42,
        n_jobs=-1
    )

    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    r2 = r2_score(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))

    print(f"\n===== 随机森林 评估结果 =====")
    print(f"R² = {r2:.4f}")
    print(f"MAE = {mae:.4f}")
    print(f"RMSE = {rmse:.4f}")

    # 特征重要性
    importance = pd.DataFrame({
        'feature': X_train.columns,
        'importance': model.feature_importances_
    }).sort_values('importance', ascending=False)

    print(f"\n特征重要性：")
    print(importance.to_string(index=False))

    # 特征重要性图
    plt.figure(figsize=(8, 5))
    plt.barh(importance['feature'][::-1], importance['importance'][::-1])
    plt.xlabel('Importance')
    plt.title('Random Forest Feature Importance')
    plt.tight_layout()
    plt.savefig('rf_importance.png', dpi=150)
    plt.close()

    return model, {'r2': r2, 'mae': mae, 'rmse': rmse}


if __name__ == "__main__":
    from xgboost_template import load_data, feature_engineering
    from sklearn.preprocessing import StandardScaler

    X, y = load_data()
    X = feature_engineering(X)

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    scaler = StandardScaler()
    X_train_scaled = pd.DataFrame(scaler.fit_transform(X_train), columns=X.columns)
    X_test_scaled = pd.DataFrame(scaler.transform(X_test), columns=X.columns)

    model, metrics = train_random_forest(X_train_scaled, X_test_scaled, y_train, y_test)
